transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c

```diff
@@ -0,0 +1,347 @@
+    pthread_mutex_t iocbLock;
+
+    jclass errorInfoClass;
+    jmethodID errorInfoConstr;
+
+    // a resuable pool of iocb
+    struct iocb ** iocb;
+    int queueSize;
+    int iocbPut;
+    int iocbGet;
+    int used;
+};
+
+static inline struct io_control * getIOControl(JNIEnv* env, jobject pointer ) {
+    struct io_control * ioControl = (struct io_control *) (*env)->GetDirectBufferAddress(env, pointer);
+    return ioControl;
+}
+
+/**
+ remove an iocb from the pool of IOCBs. Returns null if full
+*/
+static inline struct iocb * getIOCB(struct io_control * control) {
+    pthread_mutex_lock(&(control->iocbLock));
+
+    struct iocb * iocb = 0;
+
+    #ifdef DEBUG
+       fprintf (stderr, "getIOCB::used=%d, queueSize=%d, get=%d, put=%d\n", control->used, control->queueSize, control->iocbGet, control->iocbPut);
+    #endif
+
+    if (control->used < control->queueSize)
+    {
+        control->used++;
+        iocb = control->iocb[control->iocbGet++];
+
+        if (control->iocbGet >= control->queueSize)
+        {
+           control->iocbGet = 0;
+        }
+    }
+    else
+    {
+        #ifdef DEBUG
+            fprintf (stderr, "Could not find iocb\n");
+        #endif
+    }
+
+    pthread_mutex_unlock(&(control->iocbLock));
+    return iocb;
+}
+
+/**
+  Put an iocb back on the pool of IOCBs
+*/
+static inline void putIOCB(struct io_control * control, struct iocb * iocbBack) {
+    pthread_mutex_lock(&(control->iocbLock));
+
+    #ifdef DEBUG
+       fprintf (stderr, "putIOCB::used=%d, queueSize=%d, get=%d, put=%d\n", control->used, control->queueSize, control->iocbGet, control->iocbPut);
+    #endif
+
+    control->used--;
+    control->iocb[control->iocbPut++] = iocbBack;
+    if (control->iocbPut >= control->queueSize)
+    {
+       control->iocbPut = 0;
+    }
+    pthread_mutex_unlock(&(control->iocbLock));
+}
+
+static inline void * getBuffer(JNIEnv* env, jobject pointer ) {
+    return (*env)->GetDirectBufferAddress(env, pointer);
+}
+
+JNIEXPORT jobject JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_newContext(JNIEnv* env, jclass clazz, jint queueSize) {
+
+    io_context_t libaioContext;
+    int i = 0;
+
+    int res = io_queue_init(queueSize, &libaioContext);
+    if (res) {
+        // Error, so need to release whatever was done before
+        free(libaioContext);
+
+        throwRuntimeException(env, exceptionMessage("Cannot initialize queue:", -res));
+        return NULL;
+    }
+
+    struct iocb ** iocb = (struct iocb **)malloc(sizeof(struct iocb *) * queueSize);
+
+    for (i = 0; i < queueSize; i++)
+    {
+       iocb[i] = (struct iocb *)malloc(sizeof(struct iocb));
+    }
+
+    struct io_event * events = (struct io_event *)malloc(sizeof(struct io_event) * queueSize);
+
+    struct io_control * theControl = (struct io_control *) malloc(sizeof(struct io_control));
+
+    theControl->errorInfoClass = (*env)->FindClass(env, "io/netty/channel/libaio/ErrorInfo");
```

Comment:
I'm doing this once per VM load, but I think what we had there was fine since this factory / controller was supposed to be called once per VM only.

I added a hook, what actually makes me think this library should be a separate module.

or maybe you should you should rename transport-native-epoll to just native-libs or something more generic... say you could introduce even more native stuff as you go.